modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,423 @@
+package usage
+
+import (
+	"maps"
+	"math"
+	"math/bits"
+	"net/http"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/prometheus/util/strutil"
+
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+)
+
+const (
+	tenantLabel   = "tenant"
+	trackerLabel  = "tracker"
+	missingLabel  = "__missing__"
+	overflowLabel = "__overflow__"
+)
+
+type (
+	tenantLabelsFunc func(string) map[string]string
+	tenantMaxFunc    func(string) uint64
+)
+
+type bucket struct {
+	// Configuration
+	descr  *prometheus.Desc // Configuration can change over time so it is captured with the bucket.
+	labels []string
+
+	// Runtime data
+	bytes       uint64
+	lastUpdated int64
+}
+
+func (b *bucket) Inc(bytes uint64, unix int64) {
+	b.bytes += bytes
+	b.lastUpdated = unix
+}
+
+type mapping struct {
+	from string
+	to   int // Index into the values buffer
+}
+
+type tenantUsage struct {
+	series      map[uint64]*bucket
+	constLabels prometheus.Labels
+
+	// Buffers for Observe
+	dimensions map[string]string // Originally configured dimensions
+	mapping    []mapping         // Mapping from attribute => final sanitized label. Typically few values and slice is faster than map
+	sortedKeys []string          // So we can always iterate the buffer in order, this can be precomputed up front
+	buffer1    []string          // Batch-level values
+	buffer2    []string          // Span-level values
+	buffer3    []string          // Last hashed values
+	overflow   uint64
+}
+
+// GetBuffersForDimensions takes advantage of the fact that the configuration for a tracker
+// changes slowly.  Reuses buffers from the previous call when the dimensions are the same.
+func (t *tenantUsage) GetBuffersForDimensions(dimensions map[string]string) ([]mapping, []string, []string, []string) {
+	if !maps.Equal(dimensions, t.dimensions) {
```

Comment:
I wonder if we could have like a lastUpdated timestamp in the runtimeConfigOverridesManager so we could know if the runtime has changed or not

Reply:
Yes, that is more how [enterprise-logs does it](https://github.com/grafana/enterprise-logs/blob/main/pkg/enterprise/loki/usage/module.go#L108).  Where the config changes are "pushed" to the tracker, which lets us remove this check from the hot-path.   However I didn't go this route as it seemed like it would be a large change in how our overrides work.  I'm open to doing it this way in a separate PR, if this proves to be a bottleneck.